pkg/ebpf/ingress_node_firewall_loader.go

```diff
@@ -55,43 +59,295 @@ func NewIngNodeFwController() (*IngNodeFwController, error) {
+			return fmt.Errorf("lookup network iface %q: %s", ifaceName, err)
+		}
+
+		// Attach the program.
+		l, err := link.AttachXDP(link.XDPOptions{
+			Program:   objs.IngressNodeFirewallProcess,
+			Interface: iface.Index,
+		})
+		if err != nil {
+			if strings.Contains(err.Error(), xdpEBUSYErr) {
+				return fmt.Errorf("Interface %s is already attached", ifaceName)
+			}
+			return fmt.Errorf("could not attach XDP program: %s", err)
+		}
+		// Pin the XDP program.
+		lPinDir := path.Join(infc.pinPath, ifaceName+"_link")
+		if err := l.Pin(lPinDir); err != nil {
+			return fmt.Errorf("failed to pin link to pinDir %s: %s", lPinDir, err)
+		}
+		infc.links[ifaceName] = l
+		log.Printf("Attached IngressNode Firewall program to iface %q (index %d)", iface.Name, iface.Index)
+	}
+	return nil
+}
+
+// IngressNodeFwDetach detaches the eBPF program from the list of interfaces and cleans up the interfaces.
+// Additionally, it unloads all firewall rules that are associated to the interfaces.
+func (infc *IngNodeFwController) IngressNodeFwDetach(interfaceNames ...string) error {
+	// Detach from interfaces.
+	for _, interfaceName := range interfaceNames {
+		log.Printf("Detaching IngressNode Firewall program from interface %q", interfaceName)
+		if err := infc.cleanup(interfaceName); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// PurgeStaleInterfacesFromRules goes through all existing rules and purges keys for interfaces that no longer
+// exist. It first gets the list of stale interface keys (= keys for interfaces that are not part of the current
+// list of managed links). It then goes through the current ruleset and deletes all of these keys.
+func (infc *IngNodeFwController) PurgeStaleInterfacesFromRules() error {
+	// Get stale keys (= keys inside the eBPF map that do not match any managed interfaces).
+	staleKeys, err := infc.getStaleInterfaceKeys()
+	if err != nil {
+		return err
+	}
+
+	// Delete all keys that should be deleted.
+	if err := infc.purgeKeys(staleKeys); err != nil {
+		klog.Infof("Purge keys operation encountered issues, err: %q", err)
+	}
+
+	return nil
+}
+
+// GetBPFMapContent lists all existing keys and rules inside the map. Used for unit testing.
+func (infc *IngNodeFwController) GetBPFMapContent() (map[BpfLpmIpKeySt]BpfRulesValSt, error) {
+	objs := infc.objs
+
+	// Lookup all keys inside the map and find keys that are stale.
+	keysToRules := make(map[BpfLpmIpKeySt]BpfRulesValSt)
 	var key BpfLpmIpKeySt
+	var value BpfRulesValSt
+	iterator := objs.BpfMaps.IngressNodeFirewallTableMap.Iterate()
+	for iterator.Next(&key, &value) {
+		keysToRules[key] = value
+	}
+	err := iterator.Err()
+	if err != nil {
+		return nil, err
+	}
+
+	return keysToRules, nil
+}
+
+// CleaneBPFObjs closes the current objs and removes all interface pins.
+func (infc *IngNodeFwController) CleaneBPFObjs() error {
+	if err := infc.objs.Close(); err != nil {
+		return fmt.Errorf("failed to close eBPF objs err: %q", err)
+	}
+	return nil
```

Comment:
shouldn't we also RemoveAllPins() at this point ?

Reply:
I broke the original method into 3 individual chunks to make testing and moving around the code easier (also in the future). That was part of me trying to fix the detach stuff:
~~~
  179     e.log.Info("Removing all pins")
  180     if err := nodefwloader.RemoveAllPins(); err != nil {
  181         e.log.Info("Could not remove all eBPF pins", "err", err)
  182     }
  183 
  184     e.log.Info("Removing table map")
  185     if err := nodefwloader.RemoveTableMap(); err != nil {
  186         e.log.Info("Could not remove eBPF table map", "err", err)
  187     }
  188 
  189     e.log.Info("Running cleanup of eBPF objects")
  190     if err := e.c.CleaneBPFObjs(); err != nil {
  191         e.log.Info("Could not clean eBPF objects", "err", err)
  192     }
~~~

Is that here ok for you? Otherwise, I'll make them non-public and move them back into the same public method  CleaneBPFObjs:
~~~
  277 // CleaneBPFObjs closes the current objs and removes all interface pins.
  278 // FIXME: Once we figure out why links don't close / detach correctly, possibly
  279 // recombine CleaneBPFObjs / RemoveAllPins / RemoveTableMap into a single method.
  280 func (infc *IngNodeFwController) CleaneBPFObjs() error { 
  281     if err := infc.objs.Close(); err != nil { 
  282         return fmt.Errorf("failed to close eBPF objs err: %q", err) 
  283     } 
  284     return nil
  285 } 
  286 
  287 // RemoveAllPins removes all pins for XDP.
  288 func RemoveAllPins() error { 
  289     pinDir := path.Join(bpfFSPath, xdpIngressNodeFirewallProcess) 
  290     files, err := ioutil.ReadDir(pinDir) 
  291     if err != nil { 
  292         return err
  293     } 
  294 
  295     for _, file := range files { 
  296         matched, err := regexp.Match(".*"+linkSuffix+"$", []byte(file.Name())) 
  297         if err != nil { 
  298             return err
  299         } 
  300         if matched { 
  301             if err := os.Remove(path.Join(pinDir, file.Name())); err != nil { 
  302                 return err
  303             } 
  304         } 
  305     } 
  306     return nil
  307 } 
  308 
  309 // RemoveTableMap removes the ebpf table map.
  310 func RemoveTableMap() error { 
  311     pinDir := path.Join(bpfFSPath, xdpIngressNodeFirewallProcess) 
  312     return os.Remove(path.Join(pinDir, "ingress_node_firewall_table_map")) 
  313 } 
~~~

Reply:
done